File: rust/src/problems/p72.rs
pub struct Solution;

impl Solution {
    pub fn min_distance(word1: String, word2: String) -> i32 {
        let mut dp = vec![vec![0; word2.len() + 1]; word1.len() + 1];
        let w1 = word1.as_bytes();
        let w2 = word2.as_bytes();

        dp[0] = (0..word2.len() as i32 + 1).collect();

        for i in 1..=word1.len() {
            dp[i][0] = i as i32;
            for j in 1..=word2.len() {
                if w1[i - 1] == w2[j - 1] {
                    dp[i][j] = dp[i - 1][j - 1];
                } else {
                    dp[i][j] = 1 + dp[i][j - 1].min(dp[i - 1][j]).min(dp[i - 1][j - 1]);
                }
            }
        }

        dp[word1.len()][word2.len()]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn case1() {
        let word1 = "horse".to_string();
        let word2 = "ros".to_string();
        assert_eq!(3, Solution::min_distance(word1, word2));
    }

    #[test]
    fn case2() {
        let word1 = "intention".to_string();
        let word2 = "execution".to_string();
        assert_eq!(5, Solution::min_distance(word1, word2));
    }
}

File: cpp/p72.cpp
#include <cassert>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    int minDistance(string word1, string word2) {
        const int N1 = word1.size(), N2 = word2.size();
        vector<int> dp(N2 + 1, 0);
        for (int i = 0; i < N2 + 1; i++) {
            dp[i] = i;
        }
        for (int i = 1; i < N1 + 1; i++) {
            dp[0] = i - 1;
            int pre = i - 1;
            for (int j = 1; j < N2 + 1; j++) {
                int temp = dp[j];
                if (word1[i - 1] == word2[j - 1]) {
                    dp[j] = pre;
                } else {
                    dp[j] = min(min(pre + 1, dp[j - 1] + 1), dp[j] + 1);
                }
                pre = temp;
            }
        }
        return dp[N2];
    }
};

int main() {
    assert(3 == Solution().minDistance("horse", "ros"));
    assert(5 == Solution().minDistance("intention", "execution"));
    return 0;
}